Documentation/design-docs/framework-version-resolution.md

```diff
@@ -0,0 +1,290 @@
+* `3.0.0-preview4-27415-15` will roll forward to `3.0.0`.
+* `3.0.0-preview4-27415-15` will roll forward to `3.0.1-preview1-29000-0`.
+
+But it also introduces cases which are somewhat controversial:
+* On `Minor` (the default) `3.0.0` would roll forward to `3.0.1-preview1-29000-0` even if `3.0.1` is available on the machine. This behavior is typically not expected.
+* On `LatestMajor` `2.0.0` would roll forward to `3.0.1-preview1-29000-0` even when `3.0.0` is available on the machine.
+
+Pros:
+* Consistent behavior across all versions
+
+Cons:
+* Installing a pre-release version affects apps using release versions
+* Works well only if pre-release versions have similar compatibility behavior as release versions - not many breaking changes. Also expects pre-release versions to generally work well.
+
+### Alternative to new behavior
+When resolving framework reference with a pre-release version, treat all versions the same and include both release and pre-release versions in the set. This means pre-release can resolve to both release or pre-release.
+
+When resolving framework reference with a release version, prefer release versions. This means that if there's a release version on the machine which satisfies all the requirements it will be chosen, regardless of what pre-release versions are installed. Only if there's no suitable release version, then pre-release versions are also considered (and then they're treated all the same).
+
+Given the above examples this would lead to this behavior:
+* On `Minor` (the default) `3.0.0` would roll forward to `3.0.1` even if `3.0.1-preview1-29000-0` is also available on the machine.
+* On `LatestMajor` `2.0.0` would roll forward to `3.0.0` even if `3.0.1-preview1-29000-0` is also available on the machine.
+
+Pros:
+* Installing pre-release version doesn't affect apps which use release version (unless it's needed to make the app work at all)
+* Doesn't impose any implicit expectations on the quality of pre-release versions as typically pre-release would only be used when explicitly asked for.
+* Closer to what most users would expect
+
+Cons:
+* Some special cases won't work
+
+One special case which would not work:  
+Component A which asks for `2.0.0 LatestMajor` is loaded first on a machine which has `3.0.0` and also `3.1.0-preview` installed. Because it's the first in the process it will resolve the runtime according to the above rules - that is prefer release version - and thus will select `3.0.0`.  
+Later on component B is loaded which asks for `3.1.0-preview LatestMajor` (for example the one in active development). This load will fail since `3.0.0` is not enough to run this component.  
+Loading the components in reverse order (B first and then A) will work since the `3.1.0-preview` runtime will be selected.  
+Currently there's no proposed setting (env. variable) which would make this scenario work.
```

Comment:
I like the "alternative" much better than the "old" behavior for sure. I'm less sure about "alternative" vs. "new" and so I would like to have an env var that opts in to the "new" behavior if "alternative" becoimes the plan.

This alternative will at least let the following work:

1. Clean box
2. Install X.Y.N-servicing SDK
3. dotnet build netcoreappX.Y app
4. dotnet run

This will allow us to remove the hack from our automated tests.

However, a common case where we use X.Y.Z-servicing builds is by sending them to vendors to test with Visual Studio scenarios since Visual Studio will soon be patched with X.Y.Z. It occurs to me that when they are running through these scenarios, F5 is running on the X.Y.(Z-1) that came with Visual Studio and not the X.Y.Z that is about to replace it. So I would change our test steps to set this environment variable.

Reply:
I would avoid implementiung the alternative policy.